src/Report/Lcov.php

```diff
@@ -0,0 +1,161 @@
+<?php declare(strict_types=1);
+/*
+ * This file is part of phpunit/php-code-coverage.
+ *
+ * (c) Sebastian Bergmann <sebastian@phpunit.de>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+namespace SebastianBergmann\CodeCoverage\Report;
+
+use function dirname;
+use function file_put_contents;
+use function ksort;
+use function range;
+use SebastianBergmann\CodeCoverage\CodeCoverage;
+use SebastianBergmann\CodeCoverage\Directory;
+use SebastianBergmann\CodeCoverage\Driver\WriteOperationFailedException;
+use SebastianBergmann\CodeCoverage\Node\File;
+
+/**
+ * Ref: http://ltp.sourceforge.net/coverage/lcov/geninfo.1.php.
+ */
+final class Lcov
+{
+    /**
+     * @throws WriteOperationFailedException
+     */
+    public function process(CodeCoverage $coverage, ?string $target = null, ?string $name = null): string
+    {
+        $lcovLines = [];
+        $report    = $coverage->getReport();
+
+        /** @var File $file */
+        foreach ($report as $file) {
+            $tests = $this->getCoverageByTestCase($file);
+
+            foreach ($tests as $name => $test) {
+                $lcovLines[] = 'TN:' . $this->sanitizeTestName($name);
+                $lcovLines[] = 'SF:' . $file->pathAsString();
+
+                foreach ($test['lineNumbers'] as $lineNumber) {
+                    $hit         = in_array($lineNumber, $test['lineNumbersHit'], true) ? 1 : 0;
+                    $lcovLines[] = "DA:{$lineNumber},{$hit}";
+                }
+                $lcovLines[] = 'LF:' . $file->numberOfExecutableLines();
+                $lcovLines[] = 'LH:' . $file->numberOfExecutedLines();
+
+                foreach ($test['functionLineNumbers'] as $functionName => $lineNumber) {
+                    $hit         = in_array($lineNumber, $test['functionLineNumbersHit'], true) ? 1 : 0;
+                    $lcovLines[] = "FN:{$lineNumber},{$functionName}";
+                    $lcovLines[] = "FNDA:{$hit},{$functionName}";
+                }
+                $lcovLines[] = 'FNF:' . $file->numberOfMethods();
+                $lcovLines[] = 'FNH:' . $file->numberOfTestedMethods();
+
+                foreach ($test['branchLineNumbers'] as $data) {
+                    [$branchId, $lineNumber, $branchHit] = $data;
+                    $lcovLines[]                         = "BRDA:{$lineNumber},0,{$branchId},{$branchHit}";
+                }
+                $lcovLines[] = 'BRF:' . $file->numberOfExecutableBranches();
+                $lcovLines[] = 'BRH:' . $file->numberOfExecutedBranches();
+
+                $lcovLines[] = 'end_of_record';
+            }
+        }
+
+        $buffer = implode("\n", $lcovLines) . "\n";
```

Comment:
I understand this is how the other reports do it (generate in memory and then write to disk), but I'm concerned these might turn out to be very, very large files.

What's the largest project you've tested this with?

Reply:
> `fwrite` is probably a better choice? `file_put_contents` would open and close the file on every call.
> 

What do I do about returning the "buffer" at the end? Just open up the file and return the whole thing as string? :thinking:

Reply:
> What do I do about returning the "buffer" at the end? Just open up the file and return the whole thing as string?

It's only the text report where the contents are reused by the calling application 🤔. @sebastianbergmann would you be OK if nothing was returned?

Reply:
its used for the tests as well, returning an empty string would make the lcov tests fail... I would have to refactor to test it another way... :thinking: